### PR TITLE
[BugFix] Respect configured VAE dtype in LTX2 AV decoding

### PR DESCRIPTION
**Pull request by jy-song-hub**
## Motivation

SGLang already supports configuring VAE precision via vae_precision in
`python/sglang/multimodal_gen/configs/pipeline_configs/base.py`, with CLI options including fp32, fp16, and bf16 exposed at the same file.
In the current implementation of
`python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding_av.py`, the decode path parses `vae_dtype` from `server_args.pipeline_config.vae_precision`, indicating an intent to honor this configuration. However, the existing logic ignores the parsed value and unconditionally casts both the VAE and latents to `torch.bfloat16`.
This behavior creates a mismatch between the configuration surface and runtime execution, potentially violating user expectations and causing issues on devices or backends without reliable bf16 support.

This PR updates the decode path to respect the configured VAE precision and restores the VAE to its original dtype after decoding.

## Modifications
- Replace the hardcoded `torch.bfloat16` cast in `python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding_av.py` with the configured `vae_dtype` derived from `server_args.pipeline_config.vae_precision`.                                                             
- Capture the VAE's actual original dtype before decoding instead of incorrectly treating `vae_dtype` as the original dtype.               
- Restore the VAE to its true original dtype after decoding. 

## Accuracy Tests

Tested via unit test. To avoid expanding the PR surface area, the unittest is not included in this PR. For the unittest details, see the code snippet in the comment. 

## Speed Tests and Profiling

This change only fixes dtype selection and dtype restoration in the decode path. It is not intended as a performance optimization, so no dedicated speed benchmark or profiling result is included.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #26783580879](https://github.com/sgl-project/sglang/actions/runs/26783580879)
Latest PR Test (Extra): :x: [Run #26783580740](https://github.com/sgl-project/sglang/actions/runs/26783580740)

### CODE REVIEW

**Comment by jy-song-hub** [comment]
The following unit test was used to validate the correctness of this change. As the code modification is small while the test is relatively lengthy, it is included here instead of in the codebase to avoid unnecessary bloat.
```
import contextlib
import sys
import types
import unittest
from types import SimpleNamespace
from unittest.mock import Mock, patch

import torch

from sglang.multimodal_gen.runtime.pipelines_core.stages.decoding_av import (
    LTX2AVDecodingStage,
)


class _DummyVideoProcessor:
    def __init__(self, vae_scale_factor):
        self.vae_scale_factor = vae_scale_factor

    def postprocess_video(self, video, output_type="np"):
        return video


class _FakeVAE:
    def __init__(self, *, initial_dtype=torch.float32, supports_bf16=True):
        self.dtype = initial_dtype
        self.supports_bf16 = supports_bf16
        self.latents_std = torch.ones(1, dtype=initial_dtype)
        self.latents_mean = torch.zeros(1, dtype=initial_dtype)
        self.decode_input_dtype = None
        self.to_history = []

    def parameters(self):
        return iter(())

    def to(self, device=None, dtype=None, **kwargs):
        if isinstance(device, torch.dtype) and dtype is None:
            dtype = device
            device = None

        self.to_history.append((device, dtype))

        if dtype is not None:
            if dtype == torch.bfloat16 and not self.supports_bf16:
                raise RuntimeError("bfloat16 is unsupported on this backend")
            self.dtype = dtype
            self.latents_std = self.latents_std.to(dtype=dtype)
            self.latents_mean = self.latents_mean.to(dtype=dtype)

        return self

    def eval(self):
        return self

    def decode(self, latents):
        self.decode_input_dtype = latents.dtype
        return latents


class TestLTX2AVDecodingStage(unittest.TestCase):
    def _make_stage(self, vae):
        fake_diffusers = types.ModuleType("diffusers")
        fake_video_processor = types.ModuleType("diffusers.video_processor")
        fake_video_processor.VideoProcessor = _DummyVideoProcessor
        fake_diffusers.video_processor = fake_video_processor
        fake_global_server_args = SimpleNamespace(comfyui_mode=False)

        with patch.dict(
            sys.modules,
            {
                "diffusers": fake_diffusers,
                "diffusers.video_processor": fake_video_processor,
            },
        ), patch(
            "sglang.multimodal_gen.runtime.pipelines_core.stages.base.get_global_server_args",
            return_value=fake_global_server_args,
        ):
            stage = LTX2AVDecodingStage(vae, audio_vae=object(), vocoder=object())

        stage.load_model = Mock()
        stage.offload_model = Mock()
        return stage

    @staticmethod
    def _make_batch():
        return SimpleNamespace(
            latents=torch.ones((1, 1, 1, 1, 1), dtype=torch.float32),
            trajectory_timesteps=None,
            trajectory_latents=None,
            metrics=None,
        )

    @staticmethod
    def _make_server_args(vae_precision):
        return SimpleNamespace(
            disable_autocast=True,
            pipeline_config=SimpleNamespace(
                vae_precision=vae_precision,
                vae_tiling=False,
                vae_config=SimpleNamespace(arch_config=SimpleNamespace()),
                preprocess_decoding=lambda latents, *_args, **_kwargs: latents,
            ),
        )

    def test_forward_respects_configured_vae_precision_without_forcing_bf16(self):
        stage = self._make_stage(_FakeVAE(supports_bf16=False))
        batch = self._make_batch()
        server_args = self._make_server_args("fp16")

        with patch(
            "sglang.multimodal_gen.runtime.pipelines_core.stages.decoding_av.get_local_torch_device",
            return_value="cpu",
        ), patch(
            "sglang.multimodal_gen.runtime.pipelines_core.stages.decoding_av.torch.autocast",
            side_effect=lambda *args, **kwargs: contextlib.nullcontext(),
        ):
            output_batch = stage.forward(batch, server_args)

        self.assertEqual(stage.vae.decode_input_dtype, torch.float16)
        self.assertEqual(stage.vae.dtype, torch.float32)
        self.assertTrue(
            torch.equal(output_batch.output, batch.latents.to(torch.float16))
        )

    def test_forward_restores_original_vae_dtype_after_decode(self):
        stage = self._make_stage(_FakeVAE(initial_dtype=torch.float32))
        batch = self._make_batch()
        server_args = self._make_server_args("fp16")

        with patch(
            "sglang.multimodal_gen.runtime.pipelines_core.stages.decoding_av.get_local_torch_device",
            return_value="cpu",
        ), patch(
            "sglang.multimodal_gen.runtime.pipelines_core.stages.decoding_av.torch.autocast",
            side_effect=lambda *args, **kwargs: contextlib.nullcontext(),
        ):
            stage.forward(batch, server_args)

        dtype_calls = [
            dtype for _device, dtype in stage.vae.to_history if dtype is not None
        ]
        self.assertEqual(dtype_calls[-2:], [torch.float16, torch.float32])
        self.assertEqual(stage.vae.dtype, torch.float32)


if __name__ == "__main__":
    unittest.main()
```

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding_av.py [hunk 1]
```diff
@@ -37,9 +37,16 @@ def forward(self, batch: Req, server_args: ServerArgs) -> OutputBatch:
...
         ) and not server_args.disable_autocast
 
-        original_dtype = vae_dtype
-        self.vae.to(torch.bfloat16)
-        latents = latents.to(torch.bfloat16)
+        try:
+            original_dtype = self.vae.dtype
+        except AttributeError:
+            try:
+                original_dtype = next(self.vae.parameters()).dtype
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding_av.py]: why don't we use it as default, not `self.vae.dtype`?

Comment (jy-song-hub) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding_av.py]: Thank you for the suggestion. Good point!  I switched the default to the parameter dtype, which I agree is the clearer source of truth for the VAE’s actual weight precision. I kept self.vae.dtype only as a fallback for VAE-like objects that do not expose parameters(), so the path remains robust for lightweight wrappers and tests.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding_av.py [hunk 2]
```diff
@@ -51,16 +51,28 @@ def forward(self, batch: Req, server_args: ServerArgs) -> OutputBatch:
             vae_dtype != torch.float32
         ) and not server_args.disable_autocast
 
-        original_dtype = vae_dtype
         with self.use_declared_component(component_name="vae", module=self.vae) as vae:
             assert vae is not None
             self.vae = vae
             self.vae.eval()
-            latents = batch.latents.to(get_local_torch_device(), dtype=torch.bfloat16)
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding_av.py:59]: would you combine all these PRs, and help workout a way for these issues from happening again in the future?

Comment (jy-song-hub) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding_av.py:59]: Definitely. I opened a new PR to address this class of issues more systematically: https://github.com/sgl-project/sglang/pull/27088

Instead of fixing precision mismatches case by case, this PR introduces a thin shared precision-consistency layer for diffusion and multimodal generation runtime code. The goal is to make dtype handling more centralized and explicit, so that we can consistently:

- resolve runtime precision from config;
- align model/input dtype before execution;
- restore module dtype after temporary casting;
- avoid accidental hardcoded bf16/fp16 behavior;
- reduce the chance of similar precision propagation issues happening again in future model paths.

Please take a look when you get a chance. I’m happy to adjust the scope or implementation based on feedback.
